src/main/java/pl/javastart/task/services/MixPhoneContract.java

```diff
@@ -0,0 +1,104 @@
+        super(balance, smsFee, mmsFee, minuteCallFee);
+        this.remainingSms = smsAvailable;
+        this.remainingMms = mmsAvailable;
+        this.remainingMinutes = minutesAvailable;
+    }
+
+    public int getRemainingSms() {
+        return remainingSms;
+    }
+
+    protected void setRemainingSms(int remainingSms) {
+        this.remainingSms = remainingSms;
+    }
+
+    public int getRemainingMms() {
+        return remainingMms;
+    }
+
+    protected void setRemainingMms(int remainingMms) {
+        this.remainingMms = remainingMms;
+    }
+
+    public double getRemainingMinutes() {
+        return remainingMinutes;
+    }
+
+    protected void setRemainingMinutes(double remainingMinutes) {
+        this.remainingMinutes = remainingMinutes;
+    }
+
+    @Override
+    protected int availableCallSeconds(int seconds) {
+        double remainingMinutesAsSeconds = (int) (remainingMinutes * 60);
+        double allowanceSeconds = 0;
+        if (remainingMinutesAsSeconds >= seconds) { // pobiera wszystko z darmowych minut
+            callInSeconds = callInSeconds + seconds;
+            allowanceSeconds = seconds;
+            remainingMinutes -= allowanceSeconds / 60;
+            return (int) allowanceSeconds;
+        } else if (remainingMinutesAsSeconds > 0 && remainingMinutesAsSeconds < seconds) { // wybiera reszte z darmowych minut
+            remainingMinutes = 0;
+            callInSeconds = (int) (callInSeconds + remainingMinutesAsSeconds);
+            allowanceSeconds += remainingMinutesAsSeconds;
+        }
+        double feeForSecond = minuteCallFee / 60;
+        int secondsHasToBePayedFromCredit = (int) (seconds - remainingMinutesAsSeconds);
+        double feeForSecondsFromCredit = secondsHasToBePayedFromCredit * feeForSecond;
+
+        if (creditBalance >= feeForSecondsFromCredit) { // wystarczająca ilość creditu
+            creditBalance = creditBalance - feeForSecondsFromCredit;
+            callInSeconds = callInSeconds + secondsHasToBePayedFromCredit;
+            allowanceSeconds += secondsHasToBePayedFromCredit;
+            return (int) allowanceSeconds;
+        }
+
+        int creditRating = (int) (creditBalance / feeForSecond);
+
+        if (creditBalance > 0) { // dostępna tylko część creditu
+            creditBalance -= creditRating * feeForSecond;
+            callInSeconds = (callInSeconds + creditRating);
+            allowanceSeconds += creditRating;
+        }
+        return (int) allowanceSeconds;
+    }
+
+    @Override
+    protected boolean sendSms() {
+        if (remainingSms >= 1) {
+            remainingSms--;
+            smsAmount++;
+            return true;
+        }
+        return super.sendSms();
+    }
+
+    @Override
+    protected boolean sendMms() {
+        if (remainingMms >= 1) {
+            remainingMms--;
+            mmsAmount++;
+            return true;
+        } else if (getCreditBalance() >= getMmsFee()) {
+            this.sendMms();
```

Comment:
`this.sendMms();` to wywołanie metody, w której jesteśmy, chyba nie o to chodziło